### PR TITLE
Support Doctrine Cache 2

### PR DESCRIPTION
**Pull request by derrabus**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | N/A

#### Summary

This PR proposes to indicate compatibility with Doctrine Cache 2.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -34,7 +34,7 @@
     "require": {
         "php": "^7.1 || ^8",
         "ext-pdo": "*",
-        "doctrine/cache": "^1.0",
+        "doctrine/cache": "^1.0|^2.0",
```
Comment (alcaeus) [inline composer.json:37]: For code this is fine, but for tests we'll have to require `^1.11|^2.0` to avoid `--prefer-lowest` builds failing because 1.0 doesn't contain the PSR-6 adapters.

**Review by alcaeus — CHANGES_REQUESTED** [review]
doctrine/cache dependency needs to be bumped at least for require-dev, but rest LGTM

File: composer.json (hunk 1 above)
Comment (derrabus) [reply composer.json:37]: If the adapter is not there, the tests fall back to `ArrayCache`, so we're good.

Comment (alcaeus) [reply composer.json:37]: 🤦‍♂️ Of course, yes.

**Review by alcaeus — DISMISSED** [review]

**Comment by morozov** [comment]
To fix the build failure, we need to backport https://github.com/doctrine/dbal/pull/4614.

**Review by greg0ire — DISMISSED** [review]

**Review by morozov — DISMISSED** [review]

**Review by SenseException — COMMENTED** [review]
The build is currently only loading doctrine/cache 1.x on composer install. It would be great if some of them are installing 2.x to show that the tests are also working with 2.x.

**Comment by derrabus** [comment]
@SenseException I've tested with 2.0-dev locally. If you want v2 to show up in CI, someone would have to tag `2.0.0` on the Doctrine Cache repo or add a job to this repo that tests with `"minimum-stability": "dev"`.

**Comment by greg0ire** [comment]
Let's wait for 2.0.0 to happen then?

**Comment by alcaeus** [comment]
Done: https://github.com/doctrine/cache/releases/tag/2.0.0

**Comment by derrabus** [comment]
I've added small fixes to make Psalm happy, now that Doctrine Cache 2 has been tagged.

**Review by greg0ire — DISMISSED** [review]

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php [hunk 2]
```diff
@@ -402,10 +407,20 @@ public function testResultCompatibilityWhenExecutingCacheQuery(): void
...
         self::assertInstanceOf(Result::class, $result);
         self::assertInstanceOf(Driver\ResultStatement::class, $result);
     }
+
+    private function getArrayCache(): Cache
+    {
+        $cache = class_exists(DoctrineProvider::class)
+            ? DoctrineProvider::wrap(new ArrayAdapter())
+            : new ArrayCache();
+        assert($cache instanceof Cache);
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Why is this assertion necessary? Shouldn't both versions implement the expected interface?

Comment (derrabus) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Yes, but when Doctrine Cache 2 is installed, `ArrayCache` is undefined and Psalm is confused. We can remove the assertion on 3.2.x.

Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Shouldn't there be also a condition like `class_exists()`? From the static analysis PoV, the fact that the `DoctrineProvider` class doesn't exist doesn't mean that `ArrayCache` does.

So it could be something like:
```php
if (class_exists(DoctrineProvider::class)) {
    return DoctrineProvider::wrap(new ArrayAdapter());
}

if (class_exists(ArrayCache::class)) {
    return new ArrayCache();
}

self::markTestSkipped();
```

**Review by SenseException — DISMISSED** [review]

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php (hunk 2 above)
Comment (derrabus) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: I'm using your code now. Psalm seems to be less confused that way. 🙂

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php [hunk 3]
```diff
@@ -402,10 +406,23 @@ public function testResultCompatibilityWhenExecutingCacheQuery(): void
...
+    {
+        if (class_exists(DoctrineProvider::class)) {
+            return DoctrineProvider::wrap(new ArrayAdapter());
+        }
+
+        if (class_exists(ArrayCache::class)) {
+            return new ArrayCache();
+        }
+
+        self::markTestSkipped();
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: As a second thought, I think failing (`self::fail('Cannot instantiate cache backend.')`) would be a safer option (like it would fail in the original implementation).

**Review by morozov — DISMISSED** [review]

File: tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php (hunk 3 above)
Comment (derrabus) [reply tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php]: Sure.

**Review by morozov — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @derrabus !
